content/articles/hashing-in-data-structure/index.md

```diff
@@ -0,0 +1,299 @@
+---
+layout: engineering-education
+status: publish
+published: true
+url: /hashing-in-data-structure/
+title: Hashing in data structure
+description: In this article, we will explore how the hash function converts a given key to a smaller number and uses the small number as an index in a table called a hash table.
+author: anubhav-bansal
+date: 2021-06-04T00:00:00-16:00
+topics: []
+excerpt_separator: <!--more-->
+images:
+
+  - url: /hashing-in-data-structure/hero.jpg 
+    alt: hero image example image
+---
+### Hashing in Data Structure
+
+In hashing, the idea is to use a hash function that converts a given key to a smaller number and uses the small number as an index in a table called a hash table.
+
+We generate a hash for the input using the hash function and then store the element using the generated hash as the key in the hash table.
+
+![HashTable](./HashIntro.PNG)
+
+**Hash Table**: The hash table is a collection of key-value pairs. It is used when the searching or insertion of an element is required to be fast.
+
+Operation in hash function:
+
+- **Insert** - T[ h(key) ] = value;
+  - It calculates the hash, uses it as the key and stores the value in hash table.
+- **Delete** - T[ h(key) ] = NULL;
+  - It calculates the hash, resets the value stored in the hash table for that key.
+- **Search** - return T[ h(key) ];
+  - It calculates the hash, finds and returns the value stored in the hash table for that key.
+
+**Hash Collision**: When two or more inputs are mapped to the same keys as used in the hash table.
+Example: h("John") == h( "joe")
+
+A collision cannot be completely avoided but can be minimized using a 'good' hash function and a bigger table size.
+
+The chances of hash collision are less if the table size is a prime number.
+
+#### How to choose Hash Function:
+
+- An efficient hash function should be built such that the index value of the added item is distributed equally across the table.
+- An effective collision resolution technique should be created to generate an alternate index for a key whose hash index corresponds to a previously inserted position in a hash table.
+- We must select a hash algorithm that is fast to calculate.
+
+#### Characteristics of a Good Hash Function :
+
+- **Uniform Distribution**: For distribution throughout the constructed table.
+- **Fast**: The generation of hash should be very fast, and should not produce any considerable overhead.
+
+#### Collision Hashing Techniques:
+
+1. **Open Hashing (Separate Chaining)**: In this using Linked List, new nodes are added to the list, the key act as the head pointer, pointing to several nodes having different values. If any collision two elements is having same hash value we store both of them in a same linked list.
+
```

Comment:
This sentence is hard to follow. Consider rephrasing it in a way someone can understand. Also add commas where necessary